### DISCUSSION TITLE
Why is this? "columns in source table can not have any constraints. If for example any column has a NOT NULL constraint, an error will be thrown."

### DISCUSSION THREAD

**Discussion by dmschauer** [dbt-msft/dbt-sqlserver#275] [Q&A]
The [documentation ](https://dbt-msft.github.io/dbt-msft-docs/docs/dbt-sqlserver/overview/#snapshots) says "But, columns in source table can not have any constraints. If for example any column has a NOT NULL constraint, an error will be thrown." 

- Why is this? 
- Did this make life harder for anyone's project? Or is it likely? If so, how to work around it?
- And just for clarification, the quote refers to sources specified in a `sources.yml`, right?

**Comment by mikaelene** [comment]
Hi

This constraint is for snapshots only. 

By source we mean source for the snapshot. I. E. a model that the snapshot is snapshoting. 

A snapshot is a history table of another table. When I implemented this, the snapshot was created by making a clone of the source table with constraints and all. And with the "mechanics" behind how a snapshot works, it could not have constraints. If you have a primary key in the source table, it would be created in the snapshot table. But since you are storing multiple historical versions of a row. The primary key would be violated.

A workaround is to create a view without constraints an use as source for the snapshot.
